sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/sycl_ext_oneapi_bindless_images.asciidoc

```diff
@@ -0,0 +1,541 @@
+  }
+}
+
+// Image descriptor - can use the same for all images
+sycl::ext::oneapi::image_descriptor desc(
+  {width, height}, image_channel_order::rgba, image_channel_type::fp32);
+
+// Allocate each image and save the handles
+std::vector<sycl::ext::oneapi::image_mem_handle> imgAllocations;
+for (int i = 0; i < numImages; i++) {
+  // Extension: returns the handle to the allocated memory
+  auto img_mem = sycl::ext::oneapi::allocate_image(ctxt, desc);
+  if (img_mem == nullptr) {
+    std::cout << "Error allocating image!" << std::endl;
+    return 1;
+  }
+  imgAllocations.push_back(img_mem);
+}
+
+// Copy over data to device for each image
+for (int i = 0; i < numImages; i++) {
+  // Extension: copy over data to device
+  q.ext_image_copy(imgAllocations[i], dataIn.data(), desc);
+}
+
+// Wait for copy operations to finish
+q.wait();
+
+// Create the images and return the handles
+std::vector<sycl::ext::oneapi::unsampled_image_handle> imgHandles;
+for (int i = 0; i < numImages; i++) {
+  // Extension: create the image and return the handle
+  sycl::ext::oneapi::unsampled_image_handle imgHandle =
+      sycl::ext::oneapi::create_image(ctxt, imgAllocations[i], desc);
+  imgHandles.push_back(imgHandle);
+}
+
+// Cuda stores data in column-major fashion
+// SYCL deals with indexing in row-major fashion
+// Reverse output buffer dimensions and access to convert
+// the cuda column-major data back to row-major
+buffer<float, 2> buf((float *)out.data(), range<2>{height, width});
+buffer<sycl::ext::oneapi::unsampled_image_handle, 1> imgHandlesBuf(
+    imgHandles.data(), range<1>{numImages});
+q.submit([&](handler &cgh) {
+  auto outAcc =
+      buf.get_access<access_mode::write>(cgh, range<2>{height, width});
+
+  auto imgHandleAcc =
+      imgHandlesBuf.get_access<access_mode::read>(cgh, range<1>{numImages});
+
+  cgh.parallel_for<image_addition>(
+      nd_range<2>{{width, height}, {width, height}}, [=](nd_item<2> it) {
+        size_t dim0 = it.get_local_id(0);
+        size_t dim1 = it.get_local_id(1);
+
+        // Sum each image by reading their handle
+        float sum = 0;
+        for (int i = 0; i < numImages; i++) {
+          // Extension: read image data from handle
+          sum += (sycl::ext::oneapi::read_image<float4>(
+              imgHandleAcc[i], int2(dim0, dim1)))[0];
+        }
+        outAcc[id<2>{dim1, dim0}] = sum;
+      });
+});
+
+// Using image handles requires manual synchronization
+q.wait_and_throw();
+
+// Cleanup
+for (int i = 0; i < numImages; i++) {
+  sycl::ext::oneapi::destroy_image_handle(ctxt, imgHandles[i]);
+  sycl::ext::oneapi::free_image(ctxt, imgAllocations[i]);
+}
+```
+
+== Implementation notes
+
+The current DPC++ prototype only implements the proposal for the CUDA backend,
+however we are actively exploring Level Zero with SPIR-V.
+We are looking at other backend as well in order to ensure the extension can 
+work across different backends.
+
+== Issues
```

Comment:
Adding a global comment here only so the conversation can be threaded.

Several parts of this API would allow better compile-time error checking if the `unsampled_image_handle`, `sampled_image_handle`, and `image_mem_handle` types were templated on the number of dimensions.  Is this worth considering?  I think the main downside is that it becomes impossible to create an array of image handles where the handles have diverse numbers of dimensions.  I'm not sure if that is an important use case, though.  Even if it were a use case, users could instead create a `std::tuple` of image handles that have different numbers of dimensions.

To give one example of where the compile-time checking could be improved, consider `read_image` and `write_image`.  Currently, the type of `CoordT` cannot be checked at compile-time because the compiler doesn't know the number of dimensions.  Even worse, this cannot be checked at runtime either because exceptions aren't supported in device code.  Therefore, passing a `CoordT` with the wrong number of dimensions could result in a non-obvious failure that is hard to debug.  If the image handles were templated on the number of dimensions, we could check this all at compile-time.

Reply:
> Several parts of this API would allow better compile-time error checking if the `unsampled_image_handle`, `sampled_image_handle`, and `image_mem_handle` types were templated on the number of dimensions. Is this worth considering?

It's definitely worth considering. The inability to store differently dimensioned image's handles in a single container like `vector` was the reason we didn't go with more templating, and I think it would also make SYCLomatic's job a bit harder. It does provide more opportunity for compile time error checking though. We can look into this